Objects/memoryobject.c

```diff
@@ -1699,7 +1699,9 @@ unpack_single(const char *ptr, const char *fmt)
     case 'l': UNPACK_SINGLE(ld, ptr, long); goto convert_ld;
 
     /* boolean */
-    case '?': UNPACK_SINGLE(ld, ptr, _Bool); goto convert_bool;
+    // memcpy-ing values other than 0 or 1 to a _Bool variable triggers
+    // undefined behavior, so cast from char instead. See bpo-39689.
+    case '?': ld = (_Bool)*ptr; goto convert_bool;
```

Comment:
In theory this can trigger an unaligned access if `sizeof(_Bool) > 1`.  Unaligned accesses that result in bus errors are the main reason for the `memcpy()`.

Also I wonder if UB checkers can't get smart enough to figure out that e.g. ` (_Bool)*ptr == 42` and then complain.

Reply:
> In theory this can trigger an unaligned access if sizeof(_Bool) > 1. Unaligned accesses that result in bus errors are the main reason for the memcpy().

Ah! You are right, thank you. This PR won't work as is. I won't have time to work on this soon, so I'll close it in the mean time.

> Also I wonder if UB checkers can't get smart enough to figure out that e.g. (_Bool)*ptr == 42 and then complain.

I don't have access to the C standard, but from what I've read, casting to _Bool will convert non-zero values to 1. So this should be fine.

Reply:
> So this should be fine.

Yes, I misread the construct as `*(_Bool *)ptr)`, I now see that you are deliberately dereferencing the char pointer.

Hmm, the rank of `_Bool` is the smallest, but `sizeof(char)` can still be larger than `sizeof(_Bool)`.  So I wouldn't want this construct either.

Reply:
`sizeof(char)` is guaranteed to be 1. It can't be larger than `sizeof(_Bool)`.
Alignment may be a different matter (or may not; I don't know).